[lld] Add thunks for hexagon

androm3da opened the pull request:
Without thunks, programs will encounter link errors complaining that the branch target is out of range.  Thunks will extend the range of branch targets, which is a critical need for large programs.  Thunks provide this flexibility at a cost of some modest code size increase.

When configured with the maximal feature set, the hexagon port of the linux kernel would often encounter these limitations when linking with `lld`.

The relocations which will be extended by thunks are:

* R_HEX_B22_PCREL, R_HEX_{G,L}D_PLT_B22_PCREL, R_HEX_PLT_B22_PCREL relocations have a range of ± 8MiB on the baseline
* R_HEX_B15_PCREL: ±65,532 bytes
* R_HEX_B13_PCREL: ±16,380 bytes
* R_HEX_B9_PCREL: ±1,020 bytes

Fixes #149689 

Co-authored-by: Alexey Karyakin <akaryaki@quicinc.com>

androm3da commented:
For reference, here's how the Hexagon SDK's `hexagon-link` handles this interesting test case:

```
$ /local/mnt/workspace/Qualcomm/Hexagon_SDK/5.4.1.1/tools/HEXAGON_Tools/8.7.06/Tools/bin/hexagon-llvm-mc -filetype=obj -triple=hexagon-unknown-linux-musl ../llvm-project/lld/test/ELF/hexagon-thunks-packets.s -o hexagon-thunks-packets.s.tmp.o && /local/mnt/workspace/Qualcomm/Hexagon_SDK/5.4.1.1/tools/HEXAGON_Tools/8.7.06/Tools/bin/hexagon-link -o hexagon-thunks-packets.s.tmp hexagon-thunks-packets.s.tmp.o && /local/mnt/workspace/Qualcomm/Hexagon_SDK/5.4.1.1/tools/HEXAGON_Tools/8.7.06/Tools/bin/hexagon-llvm-objdump -d --print-imm-hex hexagon-thunks-packets.s.tmp

hexagon-thunks-packets.s.tmp:	file format elf32-hexagon

Disassembly of section .text:

00000000 <myfn_b>:
       0:	00 c0 9f 52	529fc000 { 	jumpr r31 } 

00000004 <main>:
       4:	28 40 00 5a	5a004028 { 	call 0x54 <trampoline_1_for_myfn_a>
       8:	00 c0 00 78	7800c000   	r0 = #0x0 } 
       c:	24 c0 00 5a	5a00c024 { 	call 0x54 <trampoline_1_for_myfn_a> } 
      10:	f8 ff ff 5b	5bfffff8 { 	call 0x0 <myfn_b> } 
      14:	f6 60 df 5d	5ddf60f6 { 	if (p0) call 0x0 <myfn_b>
      18:	20 40 20 5d	5d204020   	if (!p0) call 0x54 <trampoline_1_for_myfn_a>
      1c:	02 c1 00 f3	f300c102   	r2 = add(r0,r1) } 
      20:	1a 40 00 5d	5d00401a { 	if (p0) call 0x54 <trampoline_1_for_myfn_a>
      24:	1a 40 20 5d	5d20401a   	if (!p0) call 0x54 <trampoline_1_for_myfn_a>
      28:	02 c1 00 f3	f300c102   	r2 = add(r0,r1) } 
      2c:	14 50 80 61	61805014 { 	if (r0==#0) jump:t 0x54
      30:	02 41 00 f3	f3004102   	r2 = add(r0,r1)
      34:	54 30 41 30	30413054   	r1 = r4; 	r4 = r5 } 
      38:	0e 50 c0 61	61c0500e { 	if (r0<=#0) jump:t 0x54
      3c:	0e 40 00 12	1200400e   	p1 = cmp.eq(r0,#0x0); if (p1.new) jump:nt 0x54 <trampoline_1_for_myfn_a>
      40:	02 41 00 f3	f3004102   	r2 = add(r0,r1)
      44:	04 c0 65 70	7065c004   	r4 = r5 } 
      48:	06 c0 00 16	1600c006 { 	r0 = #0x0 ; jump 0x54 <trampoline_1_for_myfn_a> } 
      4c:	da c0 30 16	1630c0da { 	r0 = #0x0 ; jump 0x0 <myfn_b> } 
      50:	00 c0 9f 52	529fc000 { 	jumpr r31 } 

00000054 <trampoline_1_for_myfn_a>:
      54:	00 40 08 00	00084000 { 	immext(#0x800000)
      58:	10 c0 00 58	5800c010   	jump 0x80005c <myfn_a> } 
		...

0080005c <myfn_a>:
  80005c:	04 c0 00 16	1600c004 { 	r0 = #0x0 ; jump 0x800064 <trampoline_2_for_myfn_b> } 
  800060:	00 c0 9f 52	529fc000 { 	jumpr r31 } 

00800064 <trampoline_2_for_myfn_b>:
  800064:	fe 7f f7 0f	0ff77ffe { 	immext(#0xff7fff80)
  800068:	38 c0 00 58	5800c038   	jump 0x0 <myfn_b> } 
```

androm3da commented:
@quic-akaryaki has fixed the remaining test failures with this patch.  Thanks Alexey!

I'll audit it later this week and probably mark it ready for review then.

MaskRay commented on lld/ELF/Arch/Hexagon.cpp:
```diff
@@ -253,6 +260,35 @@ static uint32_t findMaskR16(uint32_t insn) {
+  default:
+    return true;
+  }
+  llvm::report_fatal_error(StringRef(
```
This is not right. `needsThunk` should reject non-branch relocations and  unknown relocations are llvm_unreachable here.

androm3da replied:
This is inconsistent, yes.  Maybe this should be `unreachable` instead?  The default case above should handle the unknown relocations and there should be no way to execute this code.

androm3da commented on lld/ELF/Relocations.cpp:
```diff
@@ -2054,18 +2054,46 @@ void ThunkCreator::mergeThunks(ArrayRef<OutputSection *> outputSections) {
+    }
+  }
+  if (ctx.arg.emachine == EM_HEXAGON) {
+    return -getHexagonPacketOffset(isec, rel);
```
fixed, thank you.

androm3da commented on lld/ELF/Thunks.cpp:
```diff
@@ -351,6 +351,21 @@ class AVRThunk : public Thunk {
+      : Thunk(ctx, dest, 0), RelOffset(rel.offset) {
+    alignment = 4;
+  }
+  uint32_t RelOffset;
```
fixed.

androm3da commented on lld/ELF/Thunks.cpp:
```diff
@@ -1365,6 +1380,39 @@ bool PPC64LongBranchThunk::isCompatibleWith(const InputSection &isec,
+  Symbol *enclosing = isec.getEnclosingSymbol(RelOffset);
+  StringRef src = enclosing ? enclosing->getName() : isec.name;
+
+  addSymbol(saver().save("__trampoline_for_" + destination.getName() +
```
I thought that it might make sense to preserve the symbol name generated by `hexagon-link` - but I don't think anything should depend on it?  In which case, I'm happy to change it.

How about  `__hexagon_trampoline_for_ ...` instead?

MaskRay replied:
Perhaps `__hexagon_thunk_` that is shorter? The symbol name should not be depended by tools, i.e. it's the tool's responsibility to adapt linker changes.

androm3da commented on lld/ELF/Thunks.cpp:
```diff
@@ -1531,6 +1579,21 @@ static Thunk *addThunkAVR(Ctx &ctx, RelType type, Symbol &s, int64_t a) {
+  case R_HEX_GD_PLT_B22_PCREL:
+    return make<HexagonThunk>(ctx, isec, rel, s);
+  default:
+    fatal("unrecognized relocation type " + toString(rel.type));
```
This has been changed to follow the pattern used by other architectures, hopefully it's satisfactory.

> ensure this indeed unreachable

Sorry, I'm not sure I understand what would be sufficient.  Ensure it by reviewing the set of relocations for which we could add a thunk?  Or by increasing test coverage here?  Or ... ?

androm3da commented on lld/test/ELF/hexagon-thunks.s:
```diff
@@ -0,0 +1,44 @@
+# RUN: llvm-mc -filetype=obj -triple=hexagon-unknown-elf %s -o %t.o
+# RUN: ld.lld %t.o -o %t
+# RUN: llvm-objdump -d %t 2>&1 | FileCheck --check-prefix=CHECK-NONPIC %s
+# RUN: llvm-mc -filetype=obj --position-independent \
```
fixed, thanks.

androm3da commented:
This PR will need to be rebased after #115925 lands.

androm3da commented:
> This PR will need to be rebased after #115925 lands.

Done.

MaskRay commented on lld/ELF/Relocations.cpp:
```diff
@@ -2031,17 +2031,44 @@ void ThunkCreator::mergeThunks(ArrayRef<OutputSection *> outputSections) {
-    return 4;
-  default:
-    return 8;
+static const uint32_t HEXAGON_MASK_END_PACKET = 3 << 14;
```
constexpr

namespace-scope const variables are of inline linkage by default, so you can omit `static`

androm3da replied:
fixed, thank you

MaskRay commented on lld/ELF/Thunks.cpp:
```diff
@@ -402,6 +402,21 @@ class AVRThunk : public Thunk {
   void addSymbols(ThunkSection &isec) override;
 };
 
+// Hexagon CPUs need thunks for <<FIXME TBD>>
```
Please fix FIXME

androm3da replied:
Ok, fixed.  Is it useful/expected to describe the range of the relocations without the thunks here like `AVRThunk` does?  If so, I'll add that here.

MaskRay commented on lld/test/ELF/hexagon-thunks.s:
```diff
@@ -0,0 +1,44 @@
+    jumpr r31
+    .size   main, .-main
+
+    .org 0x1000000
```
This creates a large output at 16MiB, which is discourated. Use a linker script like aarch64-thunk-reuse.s to assign a high address

androm3da replied:
Fixed, thanks.

MaskRay commented on lld/test/ELF/hexagon-thunks-packets.s:
```diff
@@ -0,0 +1,119 @@
+# RUN: ld.lld --pie %t.o -o %t
+# RUN: llvm-objdump -d %t 2>&1 | FileCheck --check-prefix=CHECK-PIC %s
+
+# Packets with pc-relative relocations are more interesting because
```
`## ` for non-RUN non-CHECK lines.

What is 'Packets'?

androm3da replied:
> `## ` for non-RUN non-CHECK lines.

Fixed.  Is this an `lld`-ism or should I be doing the same in other llvm-project lit test cases?

> What is 'Packets'?

In hexagon, bundles of opcodes spanning multiple instruction words are dispatched together as one.  The braces in the assembly / disassembly syntax delimit the instructions in a packet.

Would it be helpful to add this context to this test case? or somewhere else in `lld`?

nickdesaulniers replied:
> In hexagon, bundles of opcodes spanning multiple instruction words are dispatched together as one. 

VLIW bundles?

androm3da replied:
Yes, that's right - the packets that are bundled are VLIW instructions.

I used the term "bundle" in the hopes it might be a familiar one for folks who have seen the `BUNDLE` pseudo-instruction in the LLVM compiler.

MaskRay commented on lld/test/ELF/hexagon-thunks-packets.s:
```diff
@@ -0,0 +1,119 @@
+# CHECK:  Disassembly of section .text:
+
+# CHECK-NONPIC: 000200b4 <__hexagon_thunk_myfn_a_from_.text.thunk>:
+# CHECK-NONPIC: { immext(#0x1000040)
```
Use -NEXT: whenever applicable to make FileCheck errors better when things go off

androm3da replied:
fixed, thanks.

MaskRay reviewed (changes requested):
.

MaskRay commented on lld/ELF/Arch/Hexagon.cpp:
```diff
@@ -252,6 +259,34 @@ static uint32_t findMaskR16(Ctx &ctx, uint32_t insn) {
+  case llvm::ELF::R_HEX_PLT_B22_PCREL:
+  case llvm::ELF::R_HEX_GD_PLT_B22_PCREL:
+  case llvm::ELF::R_HEX_LD_PLT_B22_PCREL:
+    return llvm::isInt<22>(offset >> 2);
```
We need tests to test the thunk range. aarch64-thunk-pi.s ppc64-long-branch.s are examples to test the exact range.

androm3da replied:
I will add the necessary tests.

MaskRay commented on lld/ELF/Arch/Hexagon.cpp:
```diff
@@ -252,6 +259,34 @@ static uint32_t findMaskR16(Ctx &ctx, uint32_t insn) {
+bool Hexagon::needsThunk(RelExpr expr, RelType type, const InputFile *file,
+                         uint64_t branchAddr, const Symbol &s,
+                         int64_t a) const {
+  return !ctx.target->inBranchRange(type, branchAddr, s.getVA(ctx, a));
```
if type is unexpected, needsThunk should return false to avoid a unreachable failure in inBranchRange.

androm3da replied:
I will investigate a resolution to this issue.

androm3da replied:
fixed

androm3da committed:
[lld] Add thunks for hexagon

Co-authored-by: Alexey Karyakin <akaryaki@quicinc.com>

androm3da committed:
needsThunk()

Feedback from review

androm3da committed:
Add thunk range test cases

Feedback from review

androm3da commented:
ping @MaskRay

androm3da commented:
@MaskRay any concerns if this gets merged?  It'd be nice if we could get this in before the upcoming 21.x branch for the sake of zig & kernel.

MaskRay commented on lld/ELF/Relocations.cpp:
```diff
@@ -2129,17 +2129,44 @@ void ThunkCreator::mergeThunks(ArrayRef<OutputSection *> outputSections) {
+      return i * 4;
+  }
+}
+static int64_t getPCBias(Ctx &ctx, const InputSection &isec,
```
blank line between functions

androm3da replied:
fixed

MaskRay commented on lld/ELF/Thunks.cpp:
```diff
@@ -1692,6 +1741,24 @@ static std::unique_ptr<Thunk> addThunkAVR(Ctx &ctx, RelType type, Symbol &s,
+                                              const InputSection &isec,
+                                              Relocation &rel, Symbol &s) {
+  switch (rel.type) {
+  case R_HEX_B9_PCREL:
```
In `needsThunk` you list 7 relocation types

  case R_HEX_B22_PCREL:
  case R_HEX_PLT_B22_PCREL:
  case R_HEX_GD_PLT_B22_PCREL:
  case R_HEX_LD_PLT_B22_PCREL:
  case R_HEX_B15_PCREL:
  case R_HEX_B13_PCREL:
  case R_HEX_B9_PCREL:

Here are only 6. Please fix it and don't add `Fatal`. It should just be unreachable. ARM has Fatal because it has complex microarch thunk logic, so some erroneous code paths are technically reachable.

androm3da replied:
fixed

MaskRay commented on lld/test/ELF/hexagon-thunk-range-b22rel.s:
```diff
@@ -0,0 +1,115 @@
+  jumpr r31
+
+## Verify thunk generation for targets beyond B22_PCREL range
+# CHECK: 000200b4 <__hexagon_thunk_target_within_range_min_from_.text.thunk>:
```
Omit the address for the `<label>:` line and keep the address for the next instruction, but not subsequent instructions.

This ensures that we will get good FileCheck errors when things go wrong. And we do not need to change too many lines when changing an address.

MaskRay commented on lld/test/ELF/hexagon-thunk-range-gdplt.s:
```diff
@@ -0,0 +1,95 @@
+# REQUIRES: hexagon
+# RUN: rm -rf %t && split-file %s %t && cd %t
+# RUN: llvm-mc -filetype=obj -triple=hexagon-unknown-elf main.s -o main.o
+# RUN: ld.lld -shared main.o -o test.so
```
`a.s -o a.o`; `ld.lld -shared a.o -o a.so` (or out.so)

MaskRay commented on lld/ELF/Thunks.cpp:
```diff
@@ -1519,6 +1535,39 @@ bool PPC64LongBranchThunk::isCompatibleWith(const InputSection &isec,
+// Hexagon Target Thunks
+static uint64_t getHexagonThunkDestVA(Ctx &ctx, const Symbol &s, int64_t a) {
+  uint64_t v = s.isInPlt(ctx) ? s.getPltVA(ctx) : s.getVA(ctx, a);
+  return SignExtend64<32>(v); // FIXME: sign extend to 64-bit?
```
Why is there a FIXME?

androm3da replied:
It was...sorry!  Removed for real now in 44bfa0463657b0abf34190eb491cdbf41b3bf50d

androm3da commented on lld/ELF/Relocations.cpp:
```diff
@@ -2129,17 +2129,44 @@ void ThunkCreator::mergeThunks(ArrayRef<OutputSection *> outputSections) {
+// relocation.
+static int getHexagonPacketOffset(const InputSection &isec,
+                                  const Relocation &rel) {
+  const ArrayRef<uint8_t> SectContents = isec.content();
```
fixed

androm3da commented on lld/ELF/Relocations.cpp:
```diff
@@ -2139,17 +2139,44 @@ void ThunkCreator::mergeThunks(ArrayRef<OutputSection *> outputSections) {
+    if (i == 3 || rel.offset < (i + 1) * 4)
+      return i * 4;
+    uint32_t instWord = 0;
+    const ArrayRef<uint8_t> InstWordContents =
```
ok - will do.   And s/SectContents/data/

androm3da committed:
fixes from review

MaskRay commented on lld/test/ELF/hexagon-thunk-range-plt.s:
```diff
@@ -0,0 +1,75 @@
+# CHECK-NEXT:      jumpr r28 }
+# CHECK-NEXT:    { trap0(#0xdb) }
+
+# CHECK: 00220250 <extern_within_range@plt>:
```
As mentioned in another review, prefer omitting the address for the `<...>:` line and checking the address for the next line. This ensure good FileCheck diagnostics when things go off.

I had to adjust many tests when I made generic layout changes in the past, and testing not adopting this convention took me a lot of time to update.

androm3da committed:
Re-format test to limit changes required

Test cases re-formatted to provide assertions about behavior,
while still permitting a simpler way to update test cases when layout
changes intentionally.

androm3da commented:
@MaskRay if this makes sense to land on `main`, I'd like to see if I can cherry-pick it to `release/21.x` afterwards.

MaskRay commented on lld/ELF/Arch/Hexagon.cpp:
```diff
@@ -258,6 +265,47 @@ static uint32_t findMaskR16(Ctx &ctx, uint32_t insn) {
+  case R_HEX_B9_PCREL:
+    return !ctx.target->inBranchRange(type, branchAddr, s.getVA(ctx, a));
+  default:
+    // For unsupported relocation types, no thunk is needed
```
Delete this comment. It doesn't convey additional meaning beside the previous comment.

androm3da replied:
Deleted in 44bfa0463657b0abf34190eb491cdbf41b3bf50d

MaskRay commented on lld/test/ELF/hexagon-thunks.s:
```diff
@@ -0,0 +1,53 @@
+
+# CHECK:  Disassembly of section .text_low:
+
+# CHECK:  000200b4 <__hexagon_thunk_myfn_from_.text.thunk>:
```
Move the address from the `<...>:` line to the next line

androm3da replied:
Somehow I encountered some puzzling-to-me failures when I made this change.  I'll sort it out and push an update when I have it addressed.

androm3da committed:
remove FIXME, delete comment

androm3da committed:
Use --no-show-raw-insn and move the addrs to after the label

androm3da committed:
Fix jump/call targets

Since ac7ceb3dabfa `[Hexagon][llvm-objdump] Improve disassembly of
Hexagon bundles`, llvm-objdump includes detail about call targets, so we
should expect to see those now.

androm3da commented:
Unfortunately I had to rebase this PR in order to effectively test it after ac7ceb3dabfa landed.  I've made the changes to the test cases to handle the new output from `llvm-objdump`.
